### PR TITLE
Add PermissionOffer and PermissionOfferResponse

### PR DESCRIPTION
**Pull request by kishikawakatsumi**
Add RLMSyncPermissionOffer and RLMSyncPermissionOfferResponse.
Move `RLMSyncManagementObjectStatus` and function to `RLMSyncUtil`.

Unit tests are split another PR because tests don't work until new ROS will be released.

CC @mrackwitz @jpsim

### CODE REVIEW

**Comment by jpsim** [comment]
@kishikawakatsumi now that [Realm Object Server beta 4.8 is out](https://github.com/realm/realm-sync-services/releases/tag/v1.0.0-BETA-4.8), can you update this PR to include the tests you mentioned? Also needs a rebase.

**Comment by kishikawakatsumi** [comment]
@jpsim PermissionOffer and PermissionOfferResponse have not worked yet even beta 4.8.

https://github.com/realm/realm-sync-services/blob/master/src/node/services/permission/service.js#L27-L30

https://github.com/realm/realm-sync-services/pull/504

**Comment by kishikawakatsumi** [comment]
@jpsim @mrackwitz @bigfish24 Add unit tests for PermissionOffer and PermissionOfferResponse

**Comment by jpsim** [comment]
@kishikawakatsumi could you share steps that could help reviewers actually test this? Which version(s) of upstream dependencies should be built to try this out?

**Comment by kishikawakatsumi** [comment]
@jpsim This PR needs a new release of ROS that includes https://github.com/realm/realm-sync-services/pull/504. I've changed the state to "in progress" for now.

**Comment by jpsim** [comment]
@kishikawakatsumi can you please update this to use [ROS v1.0.0-BETA-4.14](https://github.com/realm/realm-sync-services/releases/tag/v1.0.0-BETA-4.14) that was just released? Hopefully your tests will pass with that.

Also there's a merge conflict in the changelog now.

**Comment by kishikawakatsumi** [comment]
@jpsim @mrackwitz This PR gets ready for review again. Could you review this?

This PR adds `RLMSyncPermissionOffer`, `RLMSyncPermissionOfferResponse`, `SyncPermissionOffer` and `SyncPermissionOfferResponse` to ObjC and Swift.
These model classes should be same as [realm-sync-services](https://github.com/realm/realm-sync-services/tree/master/object-server/src/node/services/permission/models).

The test confirms that a token can be obtained through `PermissionOffer`, and then `PermissionOfferResponse` works correctly using the token.

Also if `PermissionOffer` expires, the offer cannot be used. 
`PermissionOfferResponse` doesn't work with invalid tokens.

File: Realm/RLMSyncUtil.h [hunk 1]
```diff
@@ -61,4 +62,24 @@ typedef RLM_ERROR_ENUM(NSInteger, RLMSyncError, RLMSyncErrorDomain) {
...
+     performed by the object server.
+     Refer to the `statusCode` and `statusMessage` properties for more details
+     about the error.
+     */
+    RLMSyncManagementObjectStatusError,
+};
+
+/// Convert sync management object status code (nil, 0 and others) to
+/// RLMSyncManagementObjectStatus enum
+FOUNDATION_EXTERN RLMSyncManagementObjectStatus RLMSyncPermissionStatusCodeToStatus(NSNumber<RLMInt> *statusCode);
```
Comment (mrackwitz) [inline Realm/RLMSyncUtil.h]: In the style of Foundation calling struct constructors typically `NSMake*`, we could call this `RLMMakeSyncPermissionStatus`.

**Comment by kishikawakatsumi** [comment]
@mrackwitz I agreed that `id` properties should be a primary key. But on current ROS implementation, it will abort if there is a difference between ROS models and client models. To be strict, `RLMSyncPermissionChange` and `RLMSyncPermissionOffer` work fine despite changing `id` to a primary key. But if I change `RLMSyncPermissionOfferResponse`'s `id` to a primary key, ROS aborted.

```
2016-12-20T10:15:33.702Z - detail: client: Using already open Realm file: ./realm-object-server/listener/realms/c62d2e0d8ba14342b6ca2a9bb7c8edc1.realm
2016-12-20T10:15:33.703Z - debug: permission: Realm at path /4aa9ef105fff5dd6c96cc68323a6d3aa/__management has changes
2016-12-20T10:15:33.703Z - debug: permission: Opening /4aa9ef105fff5dd6c96cc68323a6d3aa/__management
2016-12-20T10:15:33.704Z - detail: permission: management-Realm of user 4aa9ef105fff5dd6c96cc68323a6d3aa has changes
2016-12-20T10:15:33.704Z - debug: permission: Found 0 unprocessed objects of PermissionChange
2016-12-20T10:15:33.715Z - debug: permission: Found 0 unprocessed objects of PermissionOffer
2016-12-20T10:15:33.716Z - debug: permission: Found 1 unprocessed objects of PermissionOfferResponse
2016-12-20T10:15:33.716Z - detail: permission: Process PermissionOfferResponse(id='E6AC8E42-0...')
2016-12-20T10:15:33.722Z - debug: permission: Opening management-Realm /e8df750a29a035504a707da71df7204d/__management
2016-12-20T10:15:33.754Z - error: permission: Error when processing PermissionOfferResponse(id='E6AC8E42-0...'): The following changes cannot be made in additive-only schema mode:
- Primary Key for class 'PermissionChange has been removed.
- Primary Key for class 'PermissionOffer has been removed.
- Primary Key for class 'PermissionOfferResponse has been removed.
2016-12-20T10:15:33.757Z - debug: permission: Closing /4aa9ef105fff5dd6c96cc68323a6d3aa/__management
libc++abi.dylib: terminating with uncaught exception of type realm::js::Exception<realm::node::Types>: Error: The following changes cannot be made in additive-only schema mode:
- Primary Key for class 'PermissionChange has been removed.
- Primary Key for class 'PermissionOffer has been removed.
- Primary Key for class 'PermissionOfferResponse has been removed.
make: *** [serve] Abort trap: 6
```

**Comment by kishikawakatsumi** [comment]
From the discussion on our Slack, we decided to accept breaking changes. Make `id` property as a primary key. eae4fdb

File: CHANGELOG.md [hunk 2]
```diff
@@ -11,7 +11,10 @@ x.x.x Release notes (yyyy-MM-dd)
 
 ### Enhancements
 
-* None.
+* Add the `RLMSyncPermissionOffer` class to allow to offer permissions to others for
```
Comment (jpsim) [inline CHANGELOG.md]: Consider combining these and rewording:

```md
* Add `SyncPermissionOffer` and `SyncPermissionOfferResponse` classes to allow
  creating and accepting permission change events to synchronized Realms between
  different users.
```

**Comment by jpsim** [comment]
Marking as "in-progress" until @kishikawakatsumi can resolve the test failures.

**Comment by kishikawakatsumi** [comment]
@jpsim This PR now ready for review again.

File: CHANGELOG.md [hunk 3]
```diff
@@ -176,7 +176,9 @@ GM seed.
 
 ### Enhancements
 
-* None.
+* Add `SyncPermissionOffer` and `SyncPermissionOfferResponse` classes to allow
```
Comment (jpsim) [inline CHANGELOG.md]: Wrong section.

File: CHANGELOG.md [hunk 4]
```diff
@@ -31,6 +31,10 @@ x.x.x Release notes (yyyy-MM-dd)
 ### Enhancements
 
 * Realm collections can now be sorted by properties over to-one relationships.
+* Add the ability to change read, write and management permissions for
```
Comment (jpsim) [inline CHANGELOG.md]: Wrong section.

Comment (kishikawakatsumi) [inline CHANGELOG.md]: Oops, I made a mistake when I resolved the conflict. Thank you.

File: CHANGELOG.md (hunk 3 above)
Comment (kishikawakatsumi) [inline CHANGELOG.md]: Thanks. I moved to the enhancements section in next version. I also add sync breaking change entry for changing primary key of `PermissionChange`

**Comment by kishikawakatsumi** [comment]
I confirmed the tests passed with `REALM_SYNC_VERSION=1.0.0-BETA-6.5` and `REALM_OBJECT_SERVER_VERSION=1.0.0-BETA-5.1-pre`.

**Comment by kishikawakatsumi** [comment]
It seems rebasing a wrong branch. I'll fix soon.

**Pull request merged**